### ISSUE TITLE
Simplified Javascript Access

### ISSUE THREAD

**Issue by robwlakes** [marcoschwartz/aREST#8]
Hi Marco,

I have been using your aREST library and admire your work. Thanks!  I was keen not to have to add any thing to my browser's capability or set up anything to do with PHP on servers etc.  That end I have worked out a way to allow Javascript to access the aREST commands using the XMLHttpRequest().  It looked ideal but threw errors as the browser was rejecting it Arduino on the basis it did not share its domain.  I have discovered that adding the line

 client.println(F("Access-Control-Allow-Origin: *"));

into the header allowed me access, and I did develop my simplified web page access.

I am a newbie to GitHub so I am not sure how to go about all this.  I would like to submit my web page if you are interested for your perusal. 

I am not sure how much this change to the header might compromise security generally (it appears to only involve the Arduino end?) or whether using the API was the answer to this problem.  I will be interest to hear your comments.

Rob Ward

File: aREST.h
.......
template <typename T>
void send_http_headers(T client){
  client.println(F("HTTP/1.1 200 OK"));
  client.println(F("Content-Type: application/json"));
  client.println(F("Access-Control-Allow-Origin: *"));
  client.println(F("Connection: close"));
  client.println();  
}

**Comment by marcoschwartz** [comment]
Old issue, headers have been added to aREST now, closing this :)

**Comment by wblaircox** [comment]
Hi Marco, sorry to did this one up. But while I see the header information has been added, I'm getting the same access error;

XMLHttpRequest cannot load https://cloud.arest.io/xxxxx. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'file://' is therefore not allowed access.

**Comment by robwlakes** [comment]
I have come across this and I believe I solved the problem, however 
Marcus did not accept it, so I only offer it as an idea.

Have a look at what I did https://github.com/robwlakes/aREST  only one 
line needed to be changed to the header.

It maybe useful to you, it may not.

Cheers, Rob

On 28/04/16 23:31, wblaircox wrote:
>
> Hi Marco, sorry to did this one up. But while I see the header 
> information has been added, I'm getting the same access error;
>
> XMLHttpRequest cannot load https://cloud.arest.io/xxxxx. No 
> 'Access-Control-Allow-Origin' header is present on the requested 
> resource. Origin 'file://' is therefore not allowed access.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/marcoschwartz/aREST/issues/8#issuecomment-215423453>
>

-- 

*Rob Ward*
Lake Tyers Beach, 3909
Lake Tyers Beach Website <http://www.laketyersbeach.net.au>
XP to XUbuntu <http://www.laketyersbeach.net.au/XP2XU.html>

**Comment by marcoschwartz** [comment]
Hi Rob, can you link to the pull request you mentioned? Can't seem to find it :)

**Comment by robwlakes** [comment]
Hi Marco,

Sorry I am not up with GITHub procedures so I think I just cloned it, 
and made a few very simple modifications but did not submit a proper 
pull request etc

Due more to the fact I thought it was a very basic addition and you 
could add the extra line or so in OK

Here is my original GITHub link. 
https://github.com/robwlakes/aREST/blob/master/README.md

and the line I added...

client.println(F("Access-Control-Allow-Origin: *"));//Allows Browsers to 
accept the returned information


It basically allows javascript and Adroid apps to access the server with 
"GET" commands or the equivalent.

I have tried to find my examples and I suspect I may have lost them in 
laptops melt down about 4 months ago.  I think I have the two main files 
attached.

Fortunately I still have my MIT App builder project for  Android that I 
developed as well that uses the change in the aRest.  I can get the 
details of that to you somehow if you are interested.

Screen shot of the Android App.

With any luck these maybe useful to you.

Cheers, Rob

On 08/05/16 04:03, marcoschwartz wrote:
>
> Hi Rob, can you link to the pull request you mentioned? Can't seem to 
> find it :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/marcoschwartz/aREST/issues/8#issuecomment-217656543>
>

-- 

*Rob Ward*
Lake Tyers Beach, 3909
Lake Tyers Beach Website <http://www.laketyersbeach.net.au>
XP to XUbuntu <http://www.laketyersbeach.net.au/XP2XU.html>
